src/Basics/Word256.hs

```diff
@@ -0,0 +1,113 @@
+{-# language BangPatterns #-}
+{-# language DataKinds #-}
+{-# language MagicHash #-}
+{-# language UnboxedTuples #-}
+
+module Basics.Word256
+  ( -- Types
+    T
+  , T#
+  , R
+    -- Lifting
+  , lift
+  , unlift
+    -- Array
+  , read#
+  , write#
+  , index#
+  , uninitialized#
+  , initialized#
+  , copy#
+  , copyMutable#
+  , set#
+  , shrink#
+  , eq#
+  , neq#
+    -- Encode
+  , shows
+  ) where
+
+import Prelude hiding (shows,minBound,maxBound)
+
+import Data.WideWord.Word256 (Word256(Word256))
+import GHC.Exts hiding (setByteArray#)
+import GHC.Word (Word64(W64#))
+
+import qualified Prelude
+import qualified GHC.Exts as Exts
+
+type T = Word256
+type T# = (# Word#, Word#, Word#, Word# #)
+type R = 'TupleRep '[ 'WordRep, 'WordRep, 'WordRep, 'WordRep ]
+
+lift :: T# -> T
+lift (# a, b, c, d #) = Word256 (W64# a) (W64# b) (W64# c) (W64# d)
+
+unlift :: T -> T#
+unlift (Word256 (W64# a) (W64# b) (W64# c) (W64# d)) = (# a, b, c, d #)
+
+eq# :: T# -> T# -> Int#
+eq# (# x1, y1, z1, w1 #) (# x2, y2, z2, w2 #) =
+  (eqWord# x1 x2) `andI#`
+  (eqWord# y1 y2) `andI#`
+  (eqWord# z1 z2) `andI#`
+  (eqWord# w1 w2)
+
+neq# :: T# -> T# -> Int#
+neq# (# x1, y1, z1, w1 #) (# x2, y2, z2, w2 #) =
+  (neWord# x1 x2) `orI#`
+  (neWord# y1 y2) `orI#`
+  (neWord# z1 z2) `orI#`
+  (neWord# w1 w2)
+
+index# :: ByteArray# -> Int# -> T#
+index# arr# i# =
+  (# Exts.indexWordArray# arr# (4# *# i#)
+  ,  Exts.indexWordArray# arr# ((4# *# i#) +# 1#)
+  ,  Exts.indexWordArray# arr# ((4# *# i#) +# 2#)
+  ,  Exts.indexWordArray# arr# ((4# *# i#) +# 3#)
+  #)
+
+read# :: MutableByteArray# s -> Int# -> State# s -> (# State# s, T# #)
+read# arr# i# s0 = case Exts.readWordArray# arr# (2# *# i#) s0 of
+  (# s1, i0 #) -> case Exts.readWordArray# arr# ((2# *# i#) +# 1#) s1 of
+    (# s2, i1 #) -> case Exts.readWordArray# arr# ((2# *# i#) +# 2#) s2 of
+      (# s3, i2 #) -> case Exts.readWordArray# arr# ((2# *# i#) +# 3#) s3 of
+        (# s4, i3 #) -> (# s4, (# i0, i1, i2, i3 #) #)
+
+write# :: MutableByteArray# s -> Int# -> T# -> State# s -> State# s
+write# arr# i# (# a, b, c, d #) s0 =
+  case Exts.writeWordArray# arr# (2# *# i#) a s0 of
```

Comment:
Same issue, multiply by four, not by two.

Reply:
See? #2

Reply:
Agreed. That would help catch these kinds of problems.